qiskit/pulse/instructions/instruction.py

```diff
@@ -56,9 +57,12 @@ def __init__(
         self._name = name
         self._hash = None
 
-        for channel in self.channels:
-            if not isinstance(channel, Channel):
-                raise PulseError(f"Expected a channel, got {channel} instead.")
+        try:
+            for channel in self.channels:
+                if not isinstance(channel, Channel):
+                    raise PulseError(f"Expected a channel, got {channel} instead.")
+        except UnassignedReferenceError:
+            pass
```

Comment:
This _looks_ surprising to me, without knowing the context - what part of the `try` block can through an `UnassignedReferenceError`?  Nothing looks like it ought to be doing any computation, which suggests to me there's performance surprises hiding in the API.

Reply:
This is a bit awkward. The issue is that the `Reference` instruction raises this error in its `channels` property. There is no reasonable value to give for `channels` for an unassigned reference, but on the other hand there is no good reason to try to look at the channels of a `ScheduleBlock` with unassigned references. The approach taken here is to raise on accessing `Reference.channels` and have `ScheduleBlock` substitute in the assigned subroutine (when it exists) in place of the `Reference` in its `channels` and `blocks` methods. So nothing outside of the `ScheduleBlock` should be working directly with a `Reference` instruction.

Perhaps there is a more elegant way to deal with this. For this specific change to `Instruction.__init__`, an alternative would be to have `Reference.channels` return an empty list until a flag is set. Then `Reference.__init__` could call `super().__init__` and then set this flag. `Reference.channels` could always return an empty list but I think raising an exception is better because any code trying to work with the channels of a `ScheduleBlock` with unassigned references is probably doing something wrong.

By the way, I don't know if it is clear reading the PR now, but the original motivation for the `Reference` instruction was to support defining a schedule like an echoed cross-resonance gate in a separate function (or even a separate package) from the one defining the single qubit x gate. The intent was that consumer would basically define the ecr and x gates and then assign x to ecr before doing anything else.

Reply:
Ah, so `channels` is a property that does a little bit of work?  I have no problems with raising the exception if there's nothing valid to return.

I personally don't like properties doing work (I much prefer having the explicit method call) because I find it's easy to introduce massive performance hits accidentally in consumer code (people assume attribute accesses are free) and if the return value is mutable, consumers often assume mutations will propagate back to the object.  But that's not my call to make, and if the work is only very slight anyway (and the return type is immutable), I wouldn't think it's a big deal.

(I honestly didn't even read the PR closely enough to know _what_ it was trying to add - I trust you and Daniel were better reviewers of that.)

Reply:
The only work `channels` is doing at the moment is raising an exception, so it shouldn't have a performance impact. I am curious what @nkanazawa1989 thinks about having it return `[]` until after calling `super().__init__()` so that we don't have to have a special case for `Reference` in the base `Instrument.__init__` method.

Reply:
This is good idea but I don't have reasonable implementation, because `Instruction` doesn't have `channels` values. They are just a part of `self._operantds` and all subclasses report channel with known index (this is to reduce parameter assignment overhead). This means we need to dynamically rewrite method.

Reply:
I came up with another implementation in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/cdefe1633a9eb392ae6c211bad1266af94098b52. I added explicit `_validate` method so that every subclass can override. Indeed there was redundant instance check in subclasses.

Reply:
cdefe16 looks nice to me (but might be the source of the new test failure?).

Reply:
Oh I didn't notice the test was also here. Actually this test is a bug and the code here will be never accepted by real hardware. The OpenPulse simulator in Aer doesn't really simulate the measurement dynamics and thus this random channel is accepted by it.